Add Tracing for SQLAlchemy and Flask-SQLAlcemy

therealryanbonham opened the pull request:
Traces calls from SQLAlchemy ORM query and session classes, and calls from Flask-SQLAlchemy.
Also patches tox.ini so that py36 is used for coverage run, as if your default python was 27 it errors.

Added a couple of basic unit test for both sqlalchemy and flask-sqlalcemy tracing. Updated README as well.   

Welcome feedback, this is my first attempt to extend X-Ray Tracing.

haotianw465 commented on aws_xray_sdk/ext/sqlalchemy/query.py:
```diff
@@ -0,0 +1,61 @@
+    pass
+
+@decorate_all_functions(xray_on_call)   
+class XRayQuery(Query):
```
Any reason you decorate functions from Query class?

therealryanbonham replied:
Query Class called by flask-sqlalchemyt, so this gets us complete traces

haotianw465 commented on aws_xray_sdk/ext/sqlalchemy/query.py:
```diff
@@ -0,0 +1,61 @@
+from __future__ import print_function
```
Unused?

therealryanbonham replied:
Removed

haotianw465 commented on setup.py:
```diff
@@ -35,7 +35,7 @@
     ],
 
-    install_requires=['jsonpickle', 'wrapt', 'requests'],
+    install_requires=['jsonpickle', 'wrapt', 'requests', 'future'],
```
Do you need future if you remove all print statement?

therealryanbonham replied:
No future will still be needed for the super() builtin
from builtins import super

therealryanbonham commented on README.md:
````diff
@@ -175,7 +175,61 @@ app.router.add_get("/", handler)
 
 web.run_app(app)
 ```
+**Add Flask middleware**
````
Yes. Removing.

therealryanbonham commented on README.md:
````diff
@@ -175,7 +175,61 @@ app.router.add_get("/", handler)
+
+```python
+from sqlalchemy.ext.declarative import declarative_base
+from sqlalchemy import *
````
No, in fact that entire import can go. I refactored the way i was doing imports and that is not needed.

therealryanbonham commented on README.md:
```diff
@@ -175,7 +175,61 @@ app.router.add_get("/", handler)
+from aws_xray_sdk.core.context import Context
+from aws_xray_sdk.ext.sqlalchemy.query import XRaySessionMaker
+
+xray_recorder.configure(service='test', sampling=False, context=Context())
```
Done

therealryanbonham commented on README.md:
```diff
@@ -175,7 +175,61 @@ app.router.add_get("/", handler)
+session = Session()
 
+xray_recorder.end_segment()
+app = Flask(__name__)
```
You have to initialize the app from Flask in order to Initialize Flask-SQLAlchemy

db = XRayFlaskSqlAlchemy(app)

 was just trying to show complete example..

haotianw465 replied:
Sounds reasonable. I'm fine with this.

therealryanbonham commented on aws_xray_sdk/ext/flask_sqlalchemy/query.py:
```diff
@@ -0,0 +1,95 @@
+from __future__ import print_function
```
Done. Was used for debugging.

therealryanbonham commented on aws_xray_sdk/ext/flask_sqlalchemy/query.py:
```diff
@@ -0,0 +1,95 @@
+from flask_sqlalchemy import SQLAlchemy, BaseQuery, _SessionSignalEvents, get_state
+from aws_xray_sdk.ext.sqlalchemy.query import XRaySession, XRayQuery
+
+def decorate_all_functions(function_decorator):
```
Flask-SQLAlchemy is a separate python package then Flask, which is why i kept them separated.  Willing to change if needed.

haotianw465 replied:
I think we can leave it for now.

therealryanbonham commented on aws_xray_sdk/ext/flask_sqlalchemy/query.py:
```diff
@@ -0,0 +1,95 @@
+            if class_name == 'sqlalchemy.orm.query':
+                for arg in args:
+                    if isinstance(arg, aws_xray_sdk.ext.sqlalchemy.query.XRayQuery):
+                        trace.put_metadata("sql", str(arg));
```
So just change this call to set_sql(). Is there any special keys for this sql metadata, or can I just the key sql as I have have here?

therealryanbonham commented on aws_xray_sdk/ext/sqlalchemy/query.py:
```diff
@@ -0,0 +1,61 @@
+from functools import wraps
+
+
+def decorate_all_functions(function_decorator):
```
Yes good point. I meant to go back and do this, and completely forgot.

haotianw465 commented:
Hi, could you take a look at http://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-sql
When you are sending data to X-Ray back-end service, if you have invalid key under subsegment.sql, your data will not be rejected, but the data will not propagate through our system. Some of the listed keys in that linked doc might be unavailable for specific sql libraries, which is OK. 

Also for the actual query recording you have to submit that change to a separate CR. We have more strict policies for reviewing code that related to query recording. And also the query is required to be sanitized.

haotianw465 commented on tests/ext/flask_sqlalchemy/test_query.py:
```diff
@@ -0,0 +1,82 @@
+    password = db.Column(db.String(255), nullable=False)
+
+
+def _search_entity(entity, name):
```
This along with find_sub could also be a helper class shared between your two query_test.py.  And since these two functions are very general they could be under tests/ root so any other unit test can utilize your contribution.

haotianw465 commented on tests/ext/flask_sqlalchemy/test_query.py:
```diff
@@ -0,0 +1,82 @@
+    return None
+
+
+def find_sub(segment, name):
```
I would recommend rename it to find_subsegment. Abbreviate on functions is not a good practice in general if it is not absolutely needed.

therealryanbonham commented:
@haotianw465  I made the change and used the set_sql to set the sanitized SQL, and added one unit test to test sanitization. That should not me an issue with sqlalchemy orm, as it passes everything via params and those are not included in the sql output i exposed. I will go ahead and comment out the set_sql and test for that, and then make second PR for that work once this first batch is accepted.

haotianw465 commented:
Hi, per doc link http://docs.aws.amazon.com/xray/latest/devguide/xray-api-segmentdocuments.html#api-segmentdocuments-sql
This is the sample sql section
```
  "sql" : {
    "url": "jdbc:postgresql://aawijb5u25wdoy.cpamxznpdoq8.us-west-2.rds.amazonaws.com:5432/ebdb",
    "preparation": "statement",
    "database_type": "PostgreSQL",
    "database_version": "9.5.4",
    "driver_version": "PostgreSQL 9.4.1211.jre7",
    "user" : "dbuser",
    "sanitized_query" : "SELECT  *  FROM  customers  WHERE  customer_id=?;"
  }
```
Could you try to get some other info and add them to the sql section as well? For an architecture that has multiple databases the info like `database_type`, `user`, `url` could be very helpful for customers. 

For sql subsegment name we have convention of using databas@host just like the one in the sample showed `"name": "ebdb@aawijb5u25wdoy.cpamxznpdoq8.us-west-2.rds.amazonaws.com"`. This makes sure when you have multiple databases the back-end won't aggregate the latency altogether.

therealryanbonham commented:
I have a separate branch started to get the set_sql working. I have it adding: url, user, database_type and sanitized_query.

https://github.com/therealryanbonham/aws-xray-sdk-python/tree/set_sql

One issue i see with the way this tracing I implemented works is it is tracing each function in sqlalchemy and thus setting a local namespace on the subsegments, so it is not creating a remove segment.. 

Thoughts? Want that as a seperate PR?

haotianw465 commented:
`trace = xray_recorder.begin_subsegment(class_name+'.'+func.__name__)` could you change the variable name from `trace` to `subsegments`? Trace is a collection of segments and subsegments. Then you can do `subsegment.namespace = 'remote'`. Database subsegment should always be remote.

For setting sql metadata could you have `url` `user` `database_type` included in this PR and a separated PR only for `sanitized_query`?

therealryanbonham commented:
The combination or using the class and function name as the subsegment name, and namespace of remote, makes a very noisy service map, as every remote call shows up as it's own object. It seems remote namespace wants the name of the subsegment to be something like the hostname you are calling.

Something like 
xray_recorder.begin_subsegment(sql['url'])
instead of 
xray_recorder.begin_subsegment(class_name+'.'+func.__name__)

The issue in that though is the trace become just a bunch of repeat named subsegment calls.. It seems like remote namespaces need a way to be named, separate from the subsegment name...  so that multiple sql functions can be traced as subsegments to a single remote namespace.. 

Something like 

xray_recorder.begin_subsegment(class_name+'.'+func.__name__, namespace=remote, namespace_name=sql['url'])

Is that supported in any way? I have tried to figure out some solution to this but am coming up blank...

haotianw465 commented:
Hi, as I pointed out in the previous thread:
For sql subsegment name we have convention of using databas@host just like the one in the sample showed `"name": "ebdb@aawijb5u25wdoy.cpamxznpdoq8.us-west-2.rds.amazonaws.com"`. This provides proper level of aggregation on service graph. 

Using url only will result in all databases on that instance aggregated to one node, and looks like the naming of your current PR result in too many nodes on service graph. 

The mentioned naming convention works well on all of our SDKs so far and we haven't heard any complain. But it really depends on if the patching code can retrieve the info about database name and endpoint. In this case looks like you can have all the info you need to construct `database@endpoint`.

haotianw465 commented:
Hi, the PR is very close to a ready state. Do you have any update on SQL subsegment naming? Or do you prefer us to take whatever you have right now and finish the work by ourselves?

therealryanbonham commented:
@haotianw465 I think this is ready for review again. It has SQL URL, user, and datatype added to the sql metadata. The sanatized_query is commented out for this PR. Traces are named by URL and have the function name as an annotation with the key "sqlalchemy".   

I added another set of helper function to util.py to search for subsegments based on annotation key/value and updated unit test to use this as well.

haotianw465 commented on aws_xray_sdk/ext/sqlalchemy/util/decerators.py:
```diff
@@ -0,0 +1,101 @@
+        if u.password is None:
+            safe_url = u.geturl()
+        else:
+            # String password from URL
```
Do you mean strip?

haotianw465 commented on aws_xray_sdk/ext/sqlalchemy/util/decerators.py:
```diff
@@ -0,0 +1,101 @@
+        if sql is not None:
+            if getattr(c._local, 'entities', None) is not None:
+                subsegment = xray_recorder.begin_subsegment(sql['url'], namespace='remote')
+                # subsegment = xray_recorder.begin_subsegment(class_name+'.'+func.__name__, )
```
Nitpick. Probably should be removed?

haotianw465 commented on tests/ext/sqlalchemy/test_query.py:
```diff
@@ -0,0 +1,70 @@
+    """ Test calling all() on get all records.
+    Verify we run the query and return the SQL as metdata"""
+    # with capsys.disabled():
+    with capsys.disabled():
```
Duplicate code on line63 and 64?
